libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m

```diff
@@ -557,15 +557,72 @@ - (void)beginUserAgentFlow {
+                if (nil == jsonError && [json isKindOfClass:[NSDictionary class]]) {
+                    NSDictionary *dict = (NSDictionary *)json;
+                    if (dict[kSFOAuthAccessToken]) {
+                        NSString *escapedString = [self URLEncodeStringFromString:approvalUrl];
+                        NSString* approvalUrl = [NSString stringWithFormat:@"%@://%@/secur/frontdoor.jsp?sid=%@&retURL=%@", self.credentials.protocol, self.credentials.instanceUrl, dict[kSFOAuthAccessToken],escapedString];
+                        [self doLoadURL:approvalUrl withCookie:YES];
+                        swapOK = YES;
+                        self.credentials.jwt = nil;
+                    }
+                }
+            }
+            if (!swapOK) {
+                [self log:SFLogLevelInfo msg:@"Fail to swap JWT for access token"];
+                [self doLoadURL:approvalUrl withCookie:NO];
+            }
+        }];
+    }
+    else {
+        [self doLoadURL:approvalUrl withCookie:NO];
+    }
+}
+
+
+- (void)swapJWTWithcompletionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandler  {
+    NSString *url = [[NSString alloc] initWithFormat:@"%@://%@%@", self.credentials.protocol,
+                     self.credentials.instanceUrl.absoluteString,
+                     kSFOAuthEndPointToken];
+    
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url] cachePolicy:NSURLRequestReloadIgnoringCacheData
+                                                       timeoutInterval:self.timeout];
+    NSString *bodyStr = [NSString stringWithFormat:@"grant_type=urn%%3Aietf%%3Aparams%%3Aoauth%%3Agrant-type%%3Ajwt-bearer&assertion=%@",self.credentials.jwt];
+    NSData *body = [bodyStr dataUsingEncoding:NSUTF8StringEncoding];
+    [request setHTTPBody:body];
+    [request setHTTPMethod:kHttpMethodPost];
+    [request setValue:kHttpPostContentType forHTTPHeaderField:kHttpHeaderContentType];
+    
+    [[self.session dataTaskWithRequest:request completionHandler:completionHandler] resume];
+}
+
+- (NSString *)URLEncodeStringFromString:(NSString *)string
+{
+    static CFStringRef charset = CFSTR("!@#$%&*()+'\";:=,/?[] ");
+    CFStringRef str = (__bridge CFStringRef)string;
+    CFStringEncoding encoding = kCFStringEncodingUTF8;
+    return (NSString *)CFBridgingRelease(CFURLCreateStringByAddingPercentEscapes(NULL, str, NULL, charset, encoding));
+}
+
+- (void)doLoadURL:(NSString*)approvalUrl  withCookie:(bool)enableCookie {
     if (self.credentials.logLevel < kSFOAuthLogLevelInfo) {
         [self log:SFLogLevelDebug format:@"SFOAuthCoordinator:beginUserAgentFlow with %@", approvalUrl];
     }
     
     NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:approvalUrl]];
-	[request setHTTPShouldHandleCookies:NO]; // don't use shared cookies
+    if(!enableCookie) [request setHTTPShouldHandleCookies:enableCookie];
```

Comment:
I'm just saying that the value of `enableCookie` will set the argument properly, without the `if` clause, whether its value is `YES` or `NO`.